keps/sig-node/3619-supplemental-groups-policy/README.md

```diff
@@ -92,37 +94,68 @@ Items marked with (R) are required *prior to targeting to a milestone / release*
+    # NOTE: "group-in-image" is not included here 
+    #       because groups defined in /etc/group should not be attached
+    #       when gids is specified in runAsGroup
+  - image: the-image-above
+    sh: ["id"] 
 ```
-Then, groups of the first process of the container will be :
-- `1000`: specified by `runAsGroup`
-- `50000`: specified by `supplementalGroups`
-- `60000`: defined in the image (<strong><em>Note: Kubernetes keep this</em></strong>)
 
-Thus, the container logs will be:
+However, the current combination with Kubernetes and major containter runtimes(at least containerd and cri-o) outputs(See [here](https://github.com/pfnet-research/strict-supplementalgroups-container-runtime/tree/reproduce-bypass-supplementalgroups) for more detailed reproduction code) includes "group-in-image" group of the first container process.
 
 ```console
-uid=1000(...) gid=1000(...) groups=1000(...),50000,60000(...)
+uid=1000(alice) gid=1000(alice) groups=1000(alice),50000(group-in-image),60000
 ```
 
-This KEP proposes a complementary field named `SupplementalGroupsPolicy` that enables users to customize to respect/ignore the group membership of the primary UID of the container defined in the container image when applying `SupplementalGroups` to the container processes. 
-
 ## Motivation
 
-As described above, `SupplementalGroups` will be just added to the groups defined in the container image for the container's primary UID. When a cluster enforces some security policy for pods that protects the value of `SupplementalGroups`, the effect of its enforcement is limited, i.e., cluster users can easily bypass the policy enforcement just by using a custom image. If such a bypass happened, it would be unexpected behavior for most cluster administrators because the enforcement is almost useless. Moreover, the bypass will cause unexpected file access permission. In some use cases, the unexpected file access permission will be a security concern. For example, using `hostPath` volumes could be a severe problem because UID/GIDs matter in accessing files/directories in the volumes.
+As described above, gids attached to the first container process is not OCI image spec compliant when `runAsGroup` is specified in `SecurityContext`.
+
+Moreover, this causes security considerations as follows. When a cluster enforces some security policy for pods that protects the value of `RunAsGroup` and `SupplementalGroups`, the effect of its enforcement is limited, i.e., cluster users can easily bypass the policy enforcement just by using a custom image. If such a bypass happened, it would be unexpected behavior for most cluster administrators because the enforcement is almost useless. Moreover, the bypass will cause unexpected file access permission. In some use cases, the unexpected file access permission will be a security concern. For example, using `hostPath` volumes could be a severe problem because UID/GIDs matter in accessing files/directories in the volumes.
```

Comment:
This paragraph makes me believe that the OCI behavior itself is incorrect.  In short:

If a pod specifies `runAsGroup` - that is authoritative, and policy engines can assume it will be enacted.

If a pod specifies `supplementalGroups` - that _should be_ authoritative (regardless of whether `runAsGroup` is specified or not), but it is not today - neither in k8s nor OCI.  Therefore, polocy engines CANNOT rely on this.

I believe (open to argument, though) that we should define make our policy "strict", and be clear that it is not OCI compliant.  If we *need* an OCI-compliant policy, we could add that as a distinct policy (this is why we don't like bool fields!).

We should sugges to OCI folks that maybe they want to tighen this up in the spec, or at least document it :)

Reply:
The problems are:

1) It seems OCI image spec doesn't actually have any affordance for supplemental groups in the image, but still expresses an opinion about the behavior of supplemental groups when converting `Config.User` ->  `process.user.additionalGids` (as opposed to letting tools define that).

2) Changing the behavior would be a breaking change, same as we are debating here, so I would expect that it would be rejected.

That said, k8s and CRI are not necessarily a strict OCI system, and we can define our semantics how we see fit.

Reply:
I've opened a corresponding issue to get runtime-spec&runc eyes on this. It seems a subtle issue. But if it's inconsistent already, then perhaps it being a "breaking" change in OCI is what is needed?

Reply:
Thank you for opening up OCI issues.

> I believe (open to argument, though) that we should define make our policy "strict", and be clear that it is not OCI compliant. If we need an OCI-compliant policy, we could add that as a distinct policy (this is why we don't like bool fields!).
> ...
> That said, k8s and CRI are not necessarily a strict OCI system, and we can define our semantics how we see fit.

How about introducing three `SupplementalGroupsPolicy`:

- `OCIcompliant`: _current_ oci compliant mode, attach groups from `/etc/group` only when `runAsGroup` (more descriptive name would be better?)
- `Lenient` (or `IncludeGroupsInContainerImage`) (Default (current)): _always_ attaches groups from `/etc/group`.
- `Strict` or (`ExcludeGroupsInContainerImage`): _never_ attach groups from `/etc/group`

Reply:
I don't think we need an OCI compliant mode (show me a use-case and I will consider it :)

How about "Merge" and "Strict" ?

Evolving from https://github.com/kubernetes/enhancements/pull/3620#discussion_r1096186213

1) If pod specifies neither `runAsGroup` nor `supplementalGroups`: 
  - use the primary group from the image (or runtime default if not specified)
  - OCI/docker read supplemental groups from the image
  - if k8s policy is "merge" (default): read supplemental groups from the image
  - if k8s policy is "strict": set no supplemental groups
  
2) If pod specifies `runAsGroup` but not `supplementalGroups`:
  - use the specified primary group
  - OCI/docker set no supplemental groups
  - if k8s policy is "merge" (default): read supplemental groups from the image
  - if k8s policy is "strict": set no supplemental groups

3) If pod specifies `supplementalGroups` but not `runAsGroup`:
  - use the primary group from the image (or runtime default if not specified)
  - OCI/docker merge the provided supplemental groups with groups from the image
  - if k8s policy is "merge" (default): merge the provided supplemental groups with groups from the image
  - if k8s policy is "strict": use only the provided supplemental groups

4) If pod specifies both `supplementalGroups` and `runAsGroup`:
  - use the specified primary group
  - OCI/docker use only the provided supplemental groups
  - if k8s policy is "merge" (default):  merge the provided supplemental groups with groups from the image
  - if k8s policy is "strict": use only the provided supplemental groups

Or put more simply: whether `runAsGroup` is specified or not has no bearing on `supplementalGroups`.

1) If pod does not specify `supplementalGroups`: 
  - if k8s policy is "merge" (default): read supplemental groups from the image
  - if k8s policy is "strict": set no supplemental groups
  
2) If pod specifies `supplementalGroups`:
  - if k8s policy is "merge" (default): merge the provided supplemental groups with groups from the image
  - if k8s policy is "strict": use only the provided supplemental groups


What do you think?

Reply:
> What do you think?

Looks good to me! I would like to confirm that your proposal is not so different than my original one except for naming, isn't it?

Reply:
renamed policy to `Merge` and `Strict`, and refined their description in [db3bcd3](https://github.com/kubernetes/enhancements/pull/3620/commits/db3bcd370eef390d766e729d90a97ac1a523b4ef). PTAL.

Reply:
renamed policies and refined their description in db3bcd3